src/AutoRest.CSharp/Common/Output/Builders/SerializationBuilder.cs

```diff
@@ -417,14 +421,30 @@ private IEnumerable<JsonPropertySerialization> GetPropertySerializationsFromBag(
         public JsonObjectSerialization BuildJsonObjectSerialization(ObjectSchema objectSchema, SchemaObjectType objectType)
         {
             var propertyBag = new SerializationPropertyBag();
-            foreach (var objectTypeLevel in objectType.EnumerateHierarchy())
+            var addToPropertyBag = (ObjectType curObj, ObjectTypeProperty curProp) =>
             {
-                foreach (var objectTypeProperty in objectTypeLevel.Properties)
+                if (curProp.SchemaProperty != null && !propertyBag.Properties.ContainsKey(curProp))
                 {
-                    if (objectTypeProperty.SchemaProperty != null)
+                    var ms = objectType.GetForMemberSerialization(curProp.Declaration.Name);
+                    if (ms == null && curObj is SchemaObjectType sot)
                     {
-                        propertyBag.Properties.Add(objectTypeProperty, objectType.GetForMemberSerialization(objectTypeProperty.Declaration.Name));
+                        ms = sot.GetForMemberSerialization(curProp.Declaration.Name);
                     }
+                    propertyBag.Properties.Add(curProp, ms);
+                }
+            };
+            foreach (var objectTypeLevel in objectType.EnumerateHierarchy())
+            {
+                foreach (var objectTypeProperty in objectTypeLevel.Properties)
+                {
+                    addToPropertyBag(objectTypeLevel, objectTypeProperty);
+                }
+            }
+            foreach (var objectTypeLevel in objectType.EnumerateHierarchy())
+            {
+                foreach (var objectTypeProperty in objectTypeLevel.SerializationConstructor.Initializers.Select(i => i.Property))
+                {
+                    addToPropertyBag(objectTypeLevel, objectTypeProperty);
                 }
```

Comment:
why here we iterate over some properties twice?

Reply:
we iterate them twice to keep the original property order in serialization code. they can be merged if we dont need to keep the order. just let me know.

Reply:
Please add a TODO item to explain that
in the meantime I still do not understand why there could be a change.
Your change should only append some new properties to the list of properties **only when** there are attributes assigning serialization metadata to the customized properties.
Nothing should happen to existing properties. I think we need to figure out why changes are happening to existing properties because this feature should change nothing on existing properties (including their order).

Reply:
the order is changed because the order in SerializatoinCustructor is different in the order or property when there is inheritance. actually, only the 2nd foreach enough, the 1st foreach is only for keeping the original order of property

Reply:
discussed offline, we should change the property directly, so no need this part change anymore